### PR TITLE
MAINT: Rework residue and residuez

### LINKED ISSUE #4562: BUG: Misc residuez issues
From https://github.com/scipy/scipy/issues/4464
1. `r, p, k = residue([-4, 8], [1, 6, 8])` should return `k == []` instead of `[0]`
2. `residuez(1, [1, -1j])` should work as well as `residuez([1], [1, -1j])` (`atleast_1d`)
3. Diagnose failures and fix:

```
r = [4, -5, 3]
p = [-1, -1, -1]
k = []
b, a = invresz(r, p, k)
b == [2, 3, 4]
a == [1, 3, 3, 1]
```

```
B = [1, 6, 6, 2]
A = [1, -(2+1j), (1+2j), -1j]
r, p, f = residuez(B, A)
r == [-2+2.5j, 7.5+7.5j, -4.5-12j]
p == [1j, 1, 1]
f == -2j
```

Double check polarities:

```
r, p, k = residuez([1, 2, 1], [1, -1, 0.3561])
r == [-0.9041 - 5.9928j, -0.9041 + 5.9928j]
p == [-0.5 - 0.3257j, -0.5000 + 0.3257j]
k == 2.8082
```

### PR DESCRIPTION
**Pull request by nmayorov**
#### What does this implement/fix?

Implementation of partial fraction expansion relied on a textbook "residue method" with fraction differentiation and all that. I decided to investigate whether there are established numerical methods for doing PFE. 

I have found a rather old paper called "Partial fractions expansion: a review of computational methodology and efficiency" with nine methods described there. And the "textbook method" was only mentioned as the one which is not recommended for computer implementation. 

So I picked another method which is relatively easy to understand and implement, doesn't have special cases and quite efficient in terms of number of required operations. Number 6 in the paper.

I've extended the documentation and included tests for `residuez`. The new implementation passes all tests from #4562, #4464 and #4561

Overall it looks like a more solid proper implementation now. The previous one was more of a literal translation of a theoretical method and not very clean or efficiently written.

@ilayn @larsoner It's not super obvious changes, but someone needs to review this. I'm happy to clarify any specifics of the code.

Closes #4562, #4561, #10929

### CODE REVIEW

**Comment by larsoner** [comment]
This all sounds good to me. Thanks for looking into the algorithm. I'm willing to trust the tests to ensure that it's working properly. And having looked just briefly at the code, it looks like overall it's a simplification, which is great. 

Failures appear real:

https://travis-ci.org/scipy/scipy/jobs/605488928

Let me know once CIs are happy (ignoring if necessary what #10993 fixes) and I'll take a deeper look.

File: scipy/signal/tests/test_signaltools.py [hunk 1]
```diff
@@ -2618,6 +2620,118 @@ def test_resiude_degenerate(self):
...
+        assert_almost_equal(p, [-2, -4])
+        assert_equal(k.size, 0)
+
+        r, p, k = residuez(0, 1)
+        assert_equal(r.size, 0)
+        assert_equal(p.size, 0)
+        assert_equal(k.size, 0)
+
+        assert_raises(ValueError, residuez, 1, 0)
+        assert_raises(ValueError, residuez, 1, [0, 1, 2, 3])
```
Comment (larsoner) [inline scipy/signal/tests/test_signaltools.py]: There is a better way to do this nowadays:
```
with pytest.raises(ValueError, match='some regex'):
    residuez(1, 0)
```
It makes it both more specific, and more obvious what degenerate case it's testing

File: scipy/signal/signaltools.py [hunk 2]
```diff
@@ -2324,9 +2324,57 @@ def invres(r, p, k, tol=1e-3, rtype='avg'):
     return b, a
 
 
+def _compute_factors(roots, multiplicity):
+    """Compute the total polynomial divided by factors for each root."""
+    suffixes = []
+    current = np.array([1])
```
Comment (larsoner) [inline scipy/signal/signaltools.py]: Maybe you need `np.array([1], roots.dtype)` or something like this? Might fix the error, not sure

**Comment by nmayorov** [comment]
> This all sounds good to me. Thanks for looking into the algorithm. I'm willing to trust the tests to ensure that it's working properly. And having looked just briefly at the code, it looks like overall it's a simplification, which is great.

The algorithm is any complicated only for repeated roots. In this case iterations are not very difficult to understand or implement either.

> Failures appear real:

There is sort of a bug in `np.polydiv`, fixed in https://github.com/numpy/numpy/pull/10473 

I worked around it by casting `b` and `a` to either `float` or `complex` datatypes. Maybe it's not an ideal approach, but it's reasonable, simple and provides some fixed accuracy of computations. 

> with pytest.raises(ValueError, match='some regex'):
>    residuez(1, 0)

Will change to that.

**Comment by larsoner** [comment]
> > with pytest.raises(ValueError, match='some regex'):
> >     residuez(1, 0)
> Will change to that.

Okay let me know when you push a commit for this and any other remaining comments and I'll look again.

**Comment by larsoner** [comment]
@Psirus can you also take a look and see if you agree this supersedes #10929 ?

**Comment by nmayorov** [comment]
@larsoner 

I think I fixed everything what necessary.

**Review by larsoner — APPROVED** [review]
LGTM +1 for merge. Let's leave it open for a couple of days in case someone else wants to look.

Feel free to ping me for a merge on Tuesday if nobody else comments.

**Comment by larsoner** [comment]
It has been a few days, so I'll go ahead and get this in. Thanks @nmayorov !

**Pull request merged**